// @angular/fire/bundles/angular-fire-functions.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/fire'), require('rxjs'), require('rxjs/operators'), require('@angular/fire/app'), require('@angular/fire/auth'), require('firebase/app'), require('rxfire/functions'), require('firebase/functions')) :
    typeof define === 'function' && define.amd ? define('@angular/fire/functions', ['exports', '@angular/core', '@angular/fire', 'rxjs', 'rxjs/operators', '@angular/fire/app', '@angular/fire/auth', 'firebase/app', 'rxfire/functions', 'firebase/functions'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.angular = global.angular || {}, global.angular.fire = global.angular.fire || {}, global.angular.fire.functions = {}), global.ng.core, global.angular.fire, global.rxjs, global.rxjs.operators, global.angular.fire.app, global.angular.fire.auth, global.app$1, global['rxfire-functions'], global['firebase-functions']));
}(this, (function (exports, i0, fire, rxjs, operators, app, auth, app$1, functions$1, functions) { 'use strict';

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () {
                            return e[k];
                        }
                    });
                }
            });
        }
        n['default'] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);

    var Functions = /** @class */ (function () {
        function Functions(functions) {
            return functions;
        }
        return Functions;
    }());
    var FUNCTIONS_PROVIDER_NAME = 'functions';
    var FunctionsInstances = /** @class */ (function () {
        function FunctionsInstances() {
            return fire.ɵgetAllInstancesOf(FUNCTIONS_PROVIDER_NAME);
        }
        return FunctionsInstances;
    }());
    var functionInstance$ = rxjs.timer(0, 300).pipe(operators.concatMap(function () { return rxjs.from(fire.ɵgetAllInstancesOf(FUNCTIONS_PROVIDER_NAME)); }), operators.distinct());

    var PROVIDED_FUNCTIONS_INSTANCES = new i0.InjectionToken('angularfire2.functions-instances');
    function defaultFunctionsInstanceFactory(provided, defaultApp) {
        var defaultAuth = fire.ɵgetDefaultInstanceOf(FUNCTIONS_PROVIDER_NAME, provided, defaultApp);
        return new Functions(defaultAuth);
    }
    function functionsInstanceFactory(fn) {
        return function (zone) {
            var functions = fire.ɵmemoizeInstance(fn, zone);
            return new Functions(functions);
        };
    }
    var FUNCTIONS_INSTANCES_PROVIDER = {
        provide: FunctionsInstances,
        deps: [
            [new i0.Optional(), PROVIDED_FUNCTIONS_INSTANCES],
        ]
    };
    var DEFAULT_FUNCTIONS_INSTANCE_PROVIDER = {
        provide: Functions,
        useFactory: defaultFunctionsInstanceFactory,
        deps: [
            [new i0.Optional(), PROVIDED_FUNCTIONS_INSTANCES],
            app.FirebaseApp,
        ]
    };
    var FunctionsModule = /** @class */ (function () {
        function FunctionsModule() {
            app$1.registerVersion('angularfire', fire.VERSION.full, 'fn');
        }
        return FunctionsModule;
    }());
    FunctionsModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0__namespace, type: FunctionsModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    FunctionsModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0__namespace, type: FunctionsModule });
    FunctionsModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0__namespace, type: FunctionsModule, providers: [
            DEFAULT_FUNCTIONS_INSTANCE_PROVIDER,
            FUNCTIONS_INSTANCES_PROVIDER,
        ] });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.3", ngImport: i0__namespace, type: FunctionsModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        providers: [
                            DEFAULT_FUNCTIONS_INSTANCE_PROVIDER,
                            FUNCTIONS_INSTANCES_PROVIDER,
                        ]
                    }]
            }], ctorParameters: function () { return []; } });
    function provideFunctions(fn) {
        return {
            ngModule: FunctionsModule,
            providers: [{
                    provide: PROVIDED_FUNCTIONS_INSTANCES,
                    useFactory: functionsInstanceFactory(fn),
                    multi: true,
                    deps: [
                        i0.NgZone,
                        fire.ɵAngularFireSchedulers,
                        app.FirebaseApps,
                        // Defensively load Auth first, if provided
                        [new i0.Optional(), auth.AuthInstances],
                    ]
                }]
        };
    }

    // DO NOT MODIFY, this file is autogenerated by tools/build.ts
    var httpsCallableData = fire.ɵzoneWrap(functions$1.httpsCallable, true);

    // DO NOT MODIFY, this file is autogenerated by tools/build.ts
    var connectFunctionsEmulator = fire.ɵzoneWrap(functions.connectFunctionsEmulator, true);
    var getFunctions = fire.ɵzoneWrap(functions.getFunctions, true);
    var httpsCallable = fire.ɵzoneWrap(functions.httpsCallable, true);

    /**
     * Generated bundle index. Do not edit.
     */

    exports.Functions = Functions;
    exports.FunctionsInstances = FunctionsInstances;
    exports.FunctionsModule = FunctionsModule;
    exports.connectFunctionsEmulator = connectFunctionsEmulator;
    exports.functionInstance$ = functionInstance$;
    exports.getFunctions = getFunctions;
    exports.httpsCallable = httpsCallable;
    exports.httpsCallableData = httpsCallableData;
    exports.provideFunctions = provideFunctions;
    Object.keys(functions).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () {
                return functions[k];
            }
        });
    });

    Object.defineProperty(exports, '__esModule', { value: true });

})));
